FIX: app args passing

mmtrt opened the pull request:
fixes opening single file or multiple from file manager. #20

mmtrt committed:
FIX: app args passing

fixes opening single file or multiple from file manager. #20

merlijn-sebrechts commented on scripts/sommelier:
```diff
@@ -217,6 +217,18 @@ start_app() {
     local exit_code="$?"
+  elif [[ -f "${ACTION_ARGS}" ]]; then
+    # Passing args to wine apps
+    for i in "${ACTION_ARGS[@]}"; do
```
I think a better approach is to put this for loop around line 202 and just edit/overwrite the ACTION_ARGS array. This way, the logic also works for apps which are launched on a virtual desktop (line 218).

mmtrt replied:
ah I didn't thought of that some how missed it.

merlijn-sebrechts commented on scripts/sommelier:
```diff
@@ -217,6 +217,18 @@ start_app() {
     fi
     "${WINELOADER}" explorer /desktop="virtual","${native_res}" "${app}" "${ACTION_ARGS[@]}"
     local exit_code="$?"
+  elif [[ -f "${ACTION_ARGS}" ]]; then
```
Note: Not entirely sure what the result is of `-f` on an array? Does it return "true" if any argument in the array is a file? Regardless, this "elif" would be gone if you move the for loop to line 202.

mmtrt replied:
didn't test that.

mmtrt replied:
yeah it skips `elif`.

merlijn-sebrechts commented on scripts/sommelier:
```diff
@@ -217,6 +217,18 @@ start_app() {
+    for i in "${ACTION_ARGS[@]}"; do
+        # Add path in Wine form (e.g. "z:/home/user/Desktop/lol.xyz")
+        if [[ -f /${i#?:} ]]; then 
+            args+=("z:${i#?:}")   # When opened through .desktop or Wine path (e.g. z:/*)
```
Does lower-case `z` also work? I thought Windows drive names were in uppercase.

mmtrt replied:
it does work fine with snaps that I had issues previously opening files.

merlijn-sebrechts commented on scripts/sommelier:
```diff
@@ -217,6 +217,18 @@ start_app() {
+    # Passing args to wine apps
+    for i in "${ACTION_ARGS[@]}"; do
+        # Add path in Wine form (e.g. "z:/home/user/Desktop/lol.xyz")
+        if [[ -f /${i#?:} ]]; then 
```
I'm assuming the initial `/` is to ensure the path is an absolute path? Do relative paths work without the `Z:` workaround? If not, can we turn all relative paths into absolute ones and then append `Z:` to all paths?

merlijn-sebrechts replied:
Also, I think `-e` would be better in this case, so that it also works if the user is opening a directory or a socket instead of a regular file.

merlijn-sebrechts commented on scripts/sommelier:
```diff
@@ -217,6 +217,18 @@ start_app() {
+        # Add path in Wine form (e.g. "z:/home/user/Desktop/lol.xyz")
+        if [[ -f /${i#?:} ]]; then 
+            args+=("z:${i#?:}")   # When opened through .desktop or Wine path (e.g. z:/*)
+        elif [[ "${i:0:1}" = "-" ]]; then
```
I prefer `"${i}" == -*` as it's a bit more clear. (also double equal signs just because we use it everywhere else in this script)

merlijn-sebrechts reviewed (changes requested):
I think checking if the file exists and appending `Z:` is a good approach. Though it's better to just edit `ACTION_ARGS` at around line 202, since we also want this fix for apps that use the virtual desktop feature of Wine.

mmtrt committed:
Update app arg passing

fix for snaps having virtual desktop, tested works fine.

updated if checks for dir & files tested both works now.

removed unnecessary if checks.

mmtrt committed:
Use winepath instead of predefined letter in path conversion

shouldn't cause issues when user have different drive letter in place of "Z".

merlijn-sebrechts commented on scripts/sommelier:
```diff
@@ -183,9 +183,19 @@ start_app() {
     . "$SNAP/sommelier/hooks/pre-start"
   fi
 
+  if [[ -e "${ACTION_ARGS}" ]]; then
```
Can you put this `if` inside of the `for` loop, so we check each argument individually? Then if the app has multiple arguments, we only run `winepath -w` for the arguments which are valid paths.

merlijn-sebrechts commented on scripts/sommelier:
```diff
@@ -183,9 +183,19 @@ start_app() {
+    for i in "${ACTION_ARGS[@]}"; do
+      args+=( "$(winepath -w "$i")" )
+    done
+    APP_ARGS=( "${args[@]}" )
```
This can simply be `ACTION_ARGS=( "${args[@]}" )`, so you don't have to change lines 198,228 and 231. Then we also don't risk on accidentally using the wrong array somewhere down the line.

merlijn-sebrechts reviewed (changes requested):
Thanks for the update! I still have two small requests, though.

mmtrt committed:
Update app arg passing

mmtrt committed:
Update app arg passing

added support for f2k enqueue feature which passes first arg as /add then other as file or dir so this handles it.

merlijn-sebrechts commented on scripts/sommelier:
```diff
@@ -183,6 +183,25 @@ start_app() {
     . "$SNAP/sommelier/hooks/pre-start"
   fi
 
+  if [[ -e "${ACTION_ARGS}" ]]; then
+    # Convert Unix paths to Windows paths.
+    for i in "${ACTION_ARGS[@]}"; do
+      if [[ -e "${i}" ]]; then 
+      args+=( "$(winepath -w "$i")" )
+      fi
+    done
+    ACTION_ARGS=( "${args[@]}" )
+  elif [[ "$ACTION_ARGS" == "/add" ]] && [[ "$(basename $RUN_EXE)" == "foobar2000.exe" ]]; then
+    # Convert Unix paths to Windows paths.
+    # Skips first arg /add in loop.
+    for i in "${ACTION_ARGS[@]:1}"; do
+      if [[ -e "${i}" ]]; then
+      args+=( "$(winepath -w "$i")" )
+      fi
+    done
+    ACTION_ARGS=( "/add" "${args[@]}" )
+  fi
+
```
I was thinking more of something like this; I think this should work for all apps, including foobar2000. Is my thinking correct?

```suggestion
  # Convert all arguments which point to existing files/dirs to Windows paths.
  for i in "${ACTION_ARGS[@]}"; do
    if [[ -e "${i}" ]]; then
      args+=( "$(winepath -w "$i")" )
    else
      args+=( "$i" )
    fi
  done
  ACTION_ARGS=( "${args[@]}" )
```

mmtrt committed:
Update scripts/sommelier

Thanks for suggested changes @galgalesh

Co-authored-by: Merlijn Sebrechts <merlijn.sebrechts@gmail.com>

mmtrt committed:
Update scripts/sommelier

merlijn-sebrechts reviewed (approved):
This is awesome! I just tested this out with Photoscape and it indeed fixes opening files from the File Manager using Photoscape!

Even better: it also fixes relative file paths not working when `SOMMELIER_KEEP_CWD` is turned off!

merlijn-sebrechts commented:
Note: this is now released to `1.0`. Apologies for the delay.
